### DISCUSSION TITLE
Support Datastores other than table/list -based data

### DISCUSSION THREAD

**Discussion by StachuDotNet** [darklang/dark#4313] [Feedback]
I'd like to manage long-lived typed data in Dark data stores that don't always fit within the current metaphor of a DB table/list.
In my head, I call them "globals" - sort of like Secrets, but: not secret, I can mutate them, and they can be of any Dark type.

Right now you can live without this feature by writing a Function that _returns_ the static data. I think it'd be worthwhile to have something separate. Maybe Secrets are a sub-part of Globals in implementation, long-term.

**Comment by pbiggar** [comment]
Can you give some examples of the kind of data you'd want to store here?

One example of data I wouldn't like to support (at least not without significant thought) is global counters.

**Comment by StachuDotNet** [comment]
quick examples:
- (string) base URL for API calls, to be prepared for the URL changing
- (date) last time [x] happened
  if you're against counters, you'd likely be against this as well
- (float) if I'm writing some special mathy program, I may want to custom encode the value of `e` as a global `2.71828`

Once we eventually no longer require manual "execution" of pure user fns, the line between a parameterless pure user function and Global may end up being in UI only, if any.

**Comment by pbiggar** [comment]
So immutable global variables are planned, which solves a majority of this.

There may be an argument to allow you to change them in the editor, such as a base url (they would still be immutable in the runtime, that is, not updatable from dark code). Though that's really just an optimization of the make-a-new-version+feature-flag+remove-old-version pattern.

**Comment by pbiggar** [comment]
Separately from immutable global variables, I think different kinds of datastore are useful. If we consider the current one to be a dict, then there's also value to 

- sets
- lists (fifo or filo)
- single atomic values (what you're talking about above)

(as well as other variations that more closely resemble caches, or that are more performant due to not supporting querying).
